# Neighborhood Segmentation & Clustering in Toronto
   ## Coursera--IBM Data Science Professional Certificate Capstone
   ## By: Abdullah M. Mustsfa
### ----------------------------------------------------------------------------------------------------------------------------------------------------

### Get Neccessary Library
First, let's get the neccessary libraries for data manipulation, analysis, modeling, and visualization.
A few packages were missing and needed to be installed.

In [1]:
#!pip install geocoder
#!pip install geopy
#!conda install -c conda-forge folium=0.5.0 --yes

In [2]:
import urllib.request #Request a web page through its URL to return its HTML content.
import requests
from bs4 import BeautifulSoup #Pasrse the HTML content into BS tree format
import pandas as pd #Tabular data manipulation in python
import numpy as np #numerical data manipulation in python
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
from sklearn.cluster import KMeans # A non-parametric clustering algorithm
import folium # Map visualisation package
import matplotlib.pyplot as plt
from matplotlib import cm, colors
%matplotlib inline

## Part 1: Retrieve a List of Postal Codes of Toronto

### Get the postal code data from the corresponding wikipedia page.
- Use urllib to request the wikipedia url and get its HTML content
- Parse the data into tree format using beautifulSoup library

In [3]:
# Define the the url of the wikipedia page
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
# open the url using urllib.request and put the HTML into the page variable
page = urllib.request.urlopen(url)
# parse the HTML from our URL into the BeautifulSoup parse tree format
soup = BeautifulSoup(page, "lxml")

### Parse the table data into a pandas dataframe
- Parse the beautifulsoup tree into corresponding postal code, borough, and neighborhood.  
- Drop postal codes with "Not assigned" Boroughs.  
- Format the table into a pandas dataframe for future manipulation.  

In [4]:
right_table=soup.find('table', class_='wikitable sortable')  # locate and retrieve the table from the soup tree
A, B , C = [] , [] , [] #initialize empty arrays
for row in right_table.findAll('tr'): #loop over the table and locate table elements 'tr' and 'td'
    cells = row.findAll('td')
    if len(cells)==3:
        # extract the table elements
        a = cells[0].find(text=True)[:-1]
        b = cells[1].find(text=True)[:-1]
        c = cells[2].find(text=True)[:-1]
        # ignore unassigned boroughs
        if b != 'Not assigned':
            A.append(a)
            B.append(b)
            C.append(c)
# Cast the arrarys into a pandas dataframe
df=pd.DataFrame({'PostalCode':A,'Borough':B,'Neighborhood':C})
df.head()

PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

#### Check for unassigned or empty neighborhoods

In [5]:
print(sum(df['Neighborhood'] == 'Not assigned'))
print(sum(df['Neighborhood'] == ''))

0
0


====> **No empty neighborhood cells were found**

#### Check for any Postal code duplicates

In [6]:
sum(df['PostalCode'].duplicated())

0

===> **No postal code duplicates** 

### Get the shape of the data

In [7]:
df.shape

(103, 3)

**===> The data contains 103 unique postal code and their associated neighborhoods and boroughs.**

## Part 2: Retrieve the corresponding latitude and longitude for each postal code

- **It wasn't possible to retrieve the data through the geocoder library.**
- **The following commented code could be used to retrieve the location data.**

In [8]:
'''
import geocoder # import geocoder
lats = []
longs = []
for postcode in df['PostalCode']:
    # initialize your variable to None
    lat_lng_coords = None
    # loop until you get the coordinates
    while(lat_lng_coords is None):
      g = geocoder.google('{}, Toronto, Ontario'.format(postcode))
      lat_lng_coords = g.latlng

    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    print(postcode, latitude, longitude)
    lats.append(latitude)
    longs.append(longitude)
'''

"\nimport geocoder # import geocoder\nlats = []\nlongs = []\nfor postcode in df['PostalCode']:\n    # initialize your variable to None\n    lat_lng_coords = None\n    # loop until you get the coordinates\n    while(lat_lng_coords is None):\n      g = geocoder.google('{}, Toronto, Ontario'.format(postcode))\n      lat_lng_coords = g.latlng\n\n    latitude = lat_lng_coords[0]\n    longitude = lat_lng_coords[1]\n    print(postcode, latitude, longitude)\n    lats.append(latitude)\n    longs.append(longitude)\n"

### Get the Geospatial data and import into a pandas dataframe.

In [9]:
geo_locs = pd.read_csv('https://cocl.us/Geospatial_data')
geo_locs.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Merge the location dataframe with the postal code dataframe

In [10]:
df_toronto = df.merge(geo_locs, left_on='PostalCode', right_on='Postal Code')
df_toronto.drop(['Postal Code'],axis=1,inplace=True) # Drop the redunant postal code info
df_toronto.head()

PostalCode           Borough                                  Neighborhood  \
0        M3A        North York                                     Parkwoods   
1        M4A        North York                              Victoria Village   
2        M5A  Downtown Toronto                    Regent Park / Harbourfront   
3        M6A        North York             Lawrence Manor / Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

## Part 3: Segmentation and Clustering of the Neighborhoods
- This is a replication of the analysis of NewYork neighborhoods.

### Retrieve the latitude and longitude of Toronto through geopy package.

In [11]:
address = 'Toronto, TO'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### Establish the Foursquare credentails.

In [12]:
CLIENT_ID = '' # your Foursquare ID
CLIENT_SECRET = '' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: UP32LRYVT2IUPMQ040YGRLE5OFYS2S2CUCVLHNSBESPFDIJM
CLIENT_SECRET:NY2TGAXSMJBZ24MQEJ5CBYF1CUK5HL4QIKISY20P1RPFJO1T


### Extract the most popular venues with every district.
- **Using Foursquare API, retreieve the most popular venues at a certain longitude and latitude.**
- **Limit the search radius to 1000 meters.**
- **Get the top 50 venues.**
- **Format the data into a pandas dataframe.**

In [13]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            100)
        
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Get the Toronto poular venues dataframe.

In [14]:
# type your answer here
toronto_venues = getNearbyVenues(names=df_toronto['Neighborhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Parkwoods              43.753259              -79.329656   
1    Parkwoods              43.753259              -79.329656   
2    Parkwoods              43.753259              -79.329656   
3    Parkwoods              43.753259              -79.329656   
4    Parkwoods              43.753259              -79.329656   

               Venue  Venue Latitude  Venue Longitude        Venue Category  
0    Allwyn's Bakery       43.759840       -79.324719  Caribbean Restaurant  
1    Brookbanks Park       43.751976       -79.332140                  Park  
2        Tim Hortons       43.760668       -79.326368                  Café  
3                A&W       43.760643       -79.326865  Fast Food Restaurant  
4  Bruno's valu-mart       43.746143       -79.324630         Grocery Store

### Distribution of venues per neighborhood

In [15]:
toronto_venues['Neighborhood'].value_counts().sample(10)

Davisville North                                                                                      100
Rouge Hill / Port Union / Highland Creek                                                                5
Don Mills                                                                                              76
New Toronto / Mimico South / Humber Bay Shores                                                         22
Kingsview Village / St. Phillips / Martin Grove Gardens / Richview Gardens                             16
Bedford Park / Lawrence Manor East                                                                     41
Woburn                                                                                                  9
Thorncliffe Park                                                                                       49
Parkdale / Roncesvalles                                                                               100
Mimico NW / The Queensway West / South of Bloo

In [16]:
toronto_venues['Neighborhood'].value_counts().describe()

count     97.000000
mean      51.051546
std       35.454011
min        3.000000
25%       20.000000
50%       41.000000
75%      100.000000
max      111.000000
Name: Neighborhood, dtype: float64

**=> The number of venues per neighborhood varies significantly.**

### Segment the neighborhood based on venues similarity.
#### Create onehot encoded vectors of the venues

In [17]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']
# move neighborhood column to the first column
fixed_columns = ['Neighborhood'] + list(set(toronto_onehot.columns) - set(['Neighborhood']))

toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  Lake  Cantonese Restaurant  Laundry Service  \
0    Parkwoods     0                     0                0   
1    Parkwoods     0                     0                0   
2    Parkwoods     0                     0                0   
3    Parkwoods     0                     0                0   
4    Parkwoods     0                     0                0   

   Japanese Restaurant  Comic Shop  French Restaurant  Women's Store  Garden  \
0                    0           0                  0              0       0   
1                    0           0                  0              0       0   
2                    0           0                  0              0       0   
3                    0           0                  0              0       0   
4                    0           0                  0              0       0   

   Asian Restaurant  ...  Gaming Cafe  Massage Studio  Hostel  \
0                 0  ...            0               0       0   
1                 0  ...            0               0       0   
2                 0  ...            0               0       0   
3                 0  ...            0               0       0   
4                 0  ...            0               0       0   

   Portuguese Restaurant  Bagel Shop  Farm  Stationery Store  Jazz Club  \
0                      0           0     0                 0          0   
1                      0           0     0                 0          0   
2                      0           0     0                 0          0   
3                      0           0     0                 0          0   
4                      0           0     0                 0          0   

   Video Store  Sandwich Place  
0            0               0  
1            0               0  
2            0               0  
3            0               0  
4            0               0  

[5 rows x 327 columns]

### Get the venues per each neighborhood using groupby
#### This dataframe can be used as a feature matrix to build neighborhood clusters

In [18]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Lake  \
0                                           Agincourt   0.0   
1                             Alderwood / Long Branch   0.0   
2   Bathurst Manor / Wilson Heights / Downsview North   0.0   
3                                     Bayview Village   0.0   
4                  Bedford Park / Lawrence Manor East   0.0   
..                                                ...   ...   
92                           Willowdale / Newtonbrook   0.0   
93                                             Woburn   0.0   
94                                   Woodbine Heights   0.0   
95                          York Mills / Silver Hills   0.0   
96                                    York Mills West   0.0   

    Cantonese Restaurant  Laundry Service  Japanese Restaurant  Comic Shop  \
0                   0.02              0.0             0.020000         0.0   
1                   0.00              0.0             0.000000         0.0   
2                   0.00              0.0             0.000000         0.0   
3                   0.00              0.0             0.133333         0.0   
4                   0.00              0.0             0.000000         0.0   
..                   ...              ...                  ...         ...   
92                  0.00              0.0             0.034483         0.0   
93                  0.00              0.0             0.000000         0.0   
94                  0.00              0.0             0.000000         0.0   
95                  0.00              0.0             0.000000         0.0   
96                  0.00              0.0             0.000000         0.0   

    French Restaurant  Women's Store  Garden  Asian Restaurant  ...  \
0            0.000000            0.0     0.0          0.020000  ...   
1            0.000000            0.0     0.0          0.000000  ...   
2            0.000000            0.0     0.0          0.000000  ...   
3            0.000000            0.0     0.0          0.000000  ...   
4            0.000000            0.0     0.0          0.000000  ...   
..                ...            ...     ...               ...  ...   
92           0.000000            0.0     0.0          0.000000  ...   
93           0.000000            0.0     0.0          0.000000  ...   
94           0.000000            0.0     0.0          0.032258  ...   
95           0.000000            0.0     0.0          0.000000  ...   
96           0.047619            0.0     0.0          0.000000  ...   

    Gaming Cafe  Massage Studio  Hostel  Portuguese Restaurant  Bagel Shop  \
0           0.0             0.0     0.0                    0.0     0.00000   
1           0.0             0.0     0.0                    0.0     0.00000   
2           0.0             0.0     0.0                    0.0     0.00000   
3           0.0             0.0     0.0                    0.0     0.00000   
4           0.0             0.0     0.0                    0.0     0.02439   
..          ...             ...     ...                    ...         ...   
92          0.0             0.0     0.0                    0.0     0.00000   
93          0.0             0.0     0.0                    0.0     0.00000   
94          0.0             0.0     0.0                    0.0     0.00000   
95          0.0             0.0     0.0                    0.0     0.00000   
96          0.0             0.0     0.0                    0.0     0.00000   

    Farm  Stationery Store  Jazz Club  Video Store  Sandwich Place  
0    0.0               0.0        0.0     0.000000        0.040000  
1    0.0               0.0        0.0     0.000000        0.040000  
2    0.0               0.0        0.0     0.032258        0.032258  
3    0.0               0.0        0.0     0.000000        0.000000  
4    0.0               0.0        0.0     0.024390        0.048780  
..   ...               ...        ...          ...             ...  
92   0.0               0.0        0.0     0.000000        0.034483  
93   

### Get Most popular venues for each neighborhood

In [19]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [20]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                          Agincourt    Chinese Restaurant   
1                            Alderwood / Long Branch              Pharmacy   
2  Bathurst Manor / Wilson Heights / Downsview North           Pizza Place   
3                                    Bayview Village         Grocery Store   
4                 Bedford Park / Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Shopping Mall        Sandwich Place           Pizza Place   
1        Discount Store           Pizza Place        Sandwich Place   
2           Coffee Shop     Convenience Store                  Bank   
3   Japanese Restaurant           Gas Station                  Bank   
4           Coffee Shop        Sandwich Place            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Caribbean Restaurant                Bakery           Coffee Shop   
1                   Pub         Grocery Store          Liquor Store   
2   Fried Chicken Joint                 Trail                 Diner   
3    Chinese Restaurant                 Trail         Shopping Mall   
4                  Bank  Fast Food Restaurant               Butcher   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Discount Store    Seafood Restaurant                   Park  
1         Shopping Mall          Intersection                    Gym  
2           Supermarket      Community Center                Dog Run  
3                  Park            Restaurant           Intersection  
4           Bridal Shop          Liquor Store       Sushi Restaurant

### Clustering using KMeans
- **Use KMeans clustering with K = 5 to cluster the neighborhoods by the venues features.**

In [21]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters,n_init=100).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 0, 0, 0, 0, 0, 0])

### Merge labels data and popular venues with other data frame

In [22]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
df_toronto_clust = df_toronto.merge(neighborhoods_venues_sorted, on='Neighborhood')

df_toronto_clust.head() # check the last columns!

PostalCode           Borough                                  Neighborhood  \
0        M3A        North York                                     Parkwoods   
1        M4A        North York                              Victoria Village   
2        M5A  Downtown Toronto                    Regent Park / Harbourfront   
3        M6A        North York             Lawrence Manor / Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656               2                  Park   
1  43.725882 -79.315572               0           Coffee Shop   
2  43.654260 -79.360636               0           Coffee Shop   
3  43.718518 -79.464763               0  Fast Food Restaurant   
4  43.662301 -79.389494               0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0              Pharmacy     Convenience Store              Bus Stop   
1           Golf Course            Boxing Gym          Hockey Arena   
2                  Café                  Park               Theater   
3            Restaurant           Coffee Shop        Clothing Store   
4      Sushi Restaurant                  Park      Ramen Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Shopping Mall    Chinese Restaurant                  Road   
1            Men's Store   Sporting Goods Shop  Gym / Fitness Center   
2                  Diner                   Pub                Bakery   
3  Vietnamese Restaurant     Accessories Store      Sushi Restaurant   
4             Restaurant   Japanese Restaurant          Burger Joint   

  8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0        Discount Store     Food & Drink Shop             Pizza Place  
1                Lounge                  Park             Pizza Place  
2        Breakfast Spot            Restaurant      Italian Restaurant  
3   Fried Chicken Joint          Dessert Shop  Furniture / Home Store  
4    Italian Restaurant             Gastropub          Ice Cream Shop

### Visualize the clustered map using Folium package

In [23]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(df_toronto_clust['Latitude'], df_toronto_clust['Longitude'], df_toronto_clust['Neighborhood'], df_toronto_clust['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

- **We can note that most of the data falls within clusters 0 and 2.**

### Display similiar neighborhood

In [24]:
df_toronto_clust.loc[df_toronto_clust['Cluster Labels'] == 0, df_toronto_clust.columns[[1] + list(range(5, df_toronto_clust.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York               0           Coffee Shop   
2    Downtown Toronto               0           Coffee Shop   
3          North York               0  Fast Food Restaurant   
4    Downtown Toronto               0           Coffee Shop   
6         Scarborough               0  Fast Food Restaurant   
7          North York               0            Restaurant   
8          North York               0            Restaurant   
10   Downtown Toronto               0           Coffee Shop   
14          East York               0                  Park   
15   Downtown Toronto               0           Coffee Shop   
16               York               0           Coffee Shop   
19       East Toronto               0                   Pub   
20   Downtown Toronto               0           Coffee Shop   
23          East York               0           Coffee Shop   
24   Downtown Toronto               0           Coffee Shop   
25   Downtown Toronto               0     Korean Restaurant   
29          East York               0           Coffee Shop   
30   Downtown Toronto               0           Coffee Shop   
31       West Toronto               0                  Café   
33         North York               0        Clothing Store   
35          East York               0           Coffee Shop   
36   Downtown Toronto               0           Coffee Shop   
37       West Toronto               0                  Café   
44       East Toronto               0      Greek Restaurant   
45   Downtown Toronto               0           Coffee Shop   
46       West Toronto               0                  Café   
49       East Toronto               0     Indian Restaurant   
50   Downtown Toronto               0           Coffee Shop   
54         North York               0     Korean Restaurant   
55       East Toronto               0           Coffee Shop   
56         North York               0    Italian Restaurant   
59        Scarborough               0          Skating Rink   
60         North York               0           Coffee Shop   
61         North York               0           Coffee Shop   
62    Central Toronto               0                  Café   
63    Central Toronto               0      Sushi Restaurant   
68    Central Toronto               0           Coffee Shop   
69    Central Toronto               0           Coffee Shop   
70       West Toronto               0                  Café   
73    Central Toronto               0           Coffee Shop   
74    Central Toronto               0                  Café   
75       West Toronto               0           Pizza Place   
76        Mississauga               0           Coffee Shop   
79    Central Toronto               0           Coffee Shop   
80   Downtown Toronto               0                  Café   
81       West Toronto               0           Coffee Shop   
83    Central Toronto               0           Coffee Shop   
84   Downtown Toronto               0                  Café   
86    Central Toronto               0           Coffee Shop   
87   Downtown Toronto               0                  Café   
88          Etobicoke               0                  Park   
92   Downtown Toronto               0           Coffee Shop   
95   Downtown Toronto               0                  Park   
96   Downtown Toronto               0           Coffee Shop   
97          Etobicoke               0           Coffee Shop   
98   Downtown Toronto               0           Coffee Shop   
99       East Toronto               0           Coffee Shop   
101         Etobicoke               0            Restaurant   

    2nd Most Common Venue          3rd Most Common Venue  \
1             Golf Course                     Boxing Gym   
2                    Café                           Park   
3              Restaurant                    Coffee Shop   
4        Sushi Restaurant                           Park

In [25]:
df_toronto_clust.loc[df_toronto_clust['Cluster Labels'] == 1, df_toronto_clust.columns[[1] + list(range(5, df_toronto_clust.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
13   Scarborough               1          Burger Joint            Playground   
100    Etobicoke               1    Italian Restaurant                  Park   

    3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
13                   Park               Breakfast Spot    Italian Restaurant   
100        Ice Cream Shop  Eastern European Restaurant  Gym / Fitness Center   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
13         Sandwich Place              Bus Line           Auto Garage   
100        Sandwich Place             BBQ Joint          Churrascaria   

    9th Most Common Venue 10th Most Common Venue  
13              BBQ Joint           Churrascaria  
100        General Travel       Ramen Restaurant

In [26]:
df_toronto_clust.loc[df_toronto_clust['Cluster Labels'] == 2, df_toronto_clust.columns[[1] + list(range(5, df_toronto_clust.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
0         North York               2                       Park   
5          Etobicoke               2                   Pharmacy   
9          East York               2                Pizza Place   
11        North York               2              Grocery Store   
12         Etobicoke               2                       Park   
17         Etobicoke               2                Coffee Shop   
18       Scarborough               2                Pizza Place   
21              York               2                       Park   
22       Scarborough               2                Coffee Shop   
26       Scarborough               2                Coffee Shop   
27        North York               2                       Park   
28        North York               2                Pizza Place   
32       Scarborough               2             Ice Cream Shop   
34        North York               2                Pizza Place   
38       Scarborough               2             Discount Store   
39        North York               2              Grocery Store   
40        North York               2                Coffee Shop   
41        North York               2                Coffee Shop   
42        North York               2                Coffee Shop   
43        North York               2                Coffee Shop   
47       Scarborough               2               Intersection   
51        North York               2                Coffee Shop   
52        North York               2          Electronics Store   
53       Scarborough               2                Pizza Place   
57              York               2     Furniture / Home Store   
58        North York               2          Convenience Store   
64              York               2                Pizza Place   
65              York               2                Coffee Shop   
66       Scarborough               2                Coffee Shop   
67        North York               2                       Park   
71         Etobicoke               2                Pizza Place   
72       Scarborough               2  Middle Eastern Restaurant   
77         Etobicoke               2                   Pharmacy   
78       Scarborough               2         Chinese Restaurant   
82       Scarborough               2                Coffee Shop   
85       Scarborough               2         Chinese Restaurant   
89         Etobicoke               2                Pizza Place   
90       Scarborough               2         Chinese Restaurant   
91  Downtown Toronto               2                       Park   
93         Etobicoke               2                   Pharmacy   

    2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
0                Pharmacy     Convenience Store                Bus Stop   
5            Skating Rink           Golf Course              Playground   
9    Fast Food Restaurant               Brewery                  Bakery   
11   Fast Food Restaurant           Pizza Place             Coffee Shop   
12            Pizza Place                 Hotel       Convenience Store   
17         Farmers Market         Shopping Mall      College Rec Center   
18   Fast Food Restaurant           Coffee Shop           Grocery Store   
21     Mexican Restaurant              Pharmacy    Fast Food Restaurant   
22                   Park     Indian Restaurant    Fast Food Restaurant   
26                 Bakery           Gas Station       Indian Restaurant   
27            Coffee Shop              Pharmacy          Sandwich Place   
28            Coffee Shop     Convenience Store                    Bank   
32         Sandwich Place         Women's Store              Restaurant   
34            Coffee Shop            Restaurant  Furniture / Home Store   
38            Coffee Shop    Chinese Restaurant           Grocery Store   
39    Japanese Restaurant           Gas Station                  

In [27]:
df_toronto_clust.loc[df_toronto_clust['Cluster Labels'] == 3, df_toronto_clust.columns[[1] + list(range(5, df_toronto_clust.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
48  North York               3                  Park                  Pool   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
48        Sandwich Place     Other Repair Shop             BBQ Joint   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
48          Churrascaria        General Travel      Ramen Restaurant   

   9th Most Common Venue 10th Most Common Venue  
48          Music School               Bus Line

In [28]:
df_toronto_clust.loc[df_toronto_clust['Cluster Labels'] == 4, df_toronto_clust.columns[[1] + list(range(5, df_toronto_clust.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
94  Etobicoke               4   Rental Car Location                 Hotel   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
94           Coffee Shop      Baseball Stadium           Auto Garage   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
94             BBQ Joint          Churrascaria        General Travel   

   9th Most Common Venue 10th Most Common Venue  
94      Ramen Restaurant           Music School

- **Some clusters contains only a single neighborhood. These could be regarded as outliers. A density based clustering would be more suitable in this case.**